### ISSUE TITLE
Possible bug in affine_autoregressive

### ISSUE THREAD

**Issue by ciaobladoo** [pyro-ppl/pyro#2503]
I think line [220](https://github.com/pyro-ppl/pyro/blob/c00ff3b8f20a28585dcbf565c645352ed2758df4/pyro/distributions/transforms/affine_autoregressive.py#L220) of affine_autoregressive.py might be wrong. Shouldn't it be like line [196](https://github.com/pyro-ppl/pyro/blob/c00ff3b8f20a28585dcbf565c645352ed2758df4/pyro/distributions/transforms/affine_autoregressive.py#L196). Otherwise it is not the same as in the naive version (line [162](https://github.com/pyro-ppl/pyro/blob/c00ff3b8f20a28585dcbf565c645352ed2758df4/pyro/distributions/transforms/affine_autoregressive.py#L162)).

In addition, I think the clamp should be applied for the inverse of stable version too since the inverse of sigmoid could also easily blow up.

**Comment by martinjankowiak** [comment]
cc @stefanwebb

**Comment by stefanwebb** [comment]
@ciaobladoo thanks for catching this bug! I've added a fix and an additional test to catch for this in the future (#2504)

I think the jury is still out on whether clamping does more good than harm in general, although happy to change it if it can be demonstrated that clamping improves optimization with some specific experiments in this case... The `torch.nn.functional.logsigmoid` should take care of some stability issues
